src/Interpreters/InterpreterParallelWithQuery.h

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include <Common/threadPoolCallbackRunner.h>
+#include <Interpreters/IInterpreter.h>
+#include <Parsers/ASTParallelWithQuery.h>
+
+
+namespace DB
+{
+
+/// Executes multiple statements in parallel.
+class InterpreterParallelWithQuery : public IInterpreter, WithContext
+{
+public:
+    InterpreterParallelWithQuery(const ASTPtr & query_, ContextPtr context_);
+    BlockIO execute() override;
+
+private:
+    void executeSubqueries(const ASTs & subqueries);
+    void executeSubquery(ASTPtr subquery, ContextMutablePtr subquery_context);
+    void waitFutures(bool throw_if_error);
+    void executeCombinedPipeline();
+
+    ASTPtr query;
+    LoggerPtr log;
+    std::unique_ptr<ThreadPool> thread_pool;
```

Comment:
This thread pool is only needed during the plan build phase, but lives the entire time of the query, while not using the running threads. Perhaps it is better to move it and runner to executeSubqueries and remove from class fields.

Reply:
I added dropping of the thread pool immediately after it stopped being used.
I'd like to keep the thread pool and the runner as class fields though because it doesn't seem very convenient to add two more arguments to `executeSubqueries()`.